storage/localstore/mode_put.go

```diff
@@ -58,119 +58,199 @@ func (db *DB) put(mode chunk.ModePut, item shed.Item) (exists bool, err error) {
-		// put to indexes: retrieve, pull
-
-		exists, err = db.retrievalDataIndex.Has(item)
-		if err != nil {
-			return exists, err
-		}
-		if !exists {
-			item.StoreTimestamp = now()
-			item.BinID, err = db.binIDs.IncInBatch(batch, uint64(db.po(item.Address)))
+		for i, ch := range chs {
+			exists, err := db.putSync(batch, binIDs, chunkToItem(ch))
 			if err != nil {
-				return false, err
+				return nil, err
+			}
+			exist[i] = exists
+			if !exists {
+				// chunk is new so, trigger pull subscription feed
+				// after the batch is successfully written
+				triggerPullFeed[db.po(ch.Address())] = struct{}{}
 			}
-			db.retrievalDataIndex.PutInBatch(batch, item)
-			db.pullIndex.PutInBatch(batch, item)
-			triggerPullFeed = true
 		}
 
 	default:
-		return false, ErrInvalidMode
+		return nil, ErrInvalidMode
+	}
+
+	for po, id := range binIDs {
+		db.binIDs.PutInBatch(batch, uint64(po), id)
 	}
 
 	err = db.incGCSizeInBatch(batch, gcSizeChange)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
 
 	err = db.shed.WriteBatch(batch)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	if triggerPullFeed {
-		db.triggerPullSubscriptions(db.po(item.Address))
+	for po := range triggerPullFeed {
+		db.triggerPullSubscriptions(po)
 	}
 	if triggerPushFeed {
 		db.triggerPushSubscriptions()
 	}
-	return exists, nil
+	return exist, nil
+}
+
+// putRequest adds an Item to the batch by updating required indexes:
+//  - put to indexes: retrieve, gc
+//  - it does not enter the syncpool
+// The batch can be written to the database.
+// Provided batch and binID map are updated.
+func (db *DB) putRequest(batch *leveldb.Batch, binIDs map[uint8]uint64, item shed.Item) (exists bool, gcSizeChange int64, err error) {
+	// check if the chunk already is in the database
+	// as gc index is updated
+	i, err := db.retrievalAccessIndex.Get(item)
+	switch err {
+	case nil:
+		exists = true
+		item.AccessTimestamp = i.AccessTimestamp
+	case leveldb.ErrNotFound:
+		exists = false
+		// no chunk accesses
+	default:
+		return false, 0, err
+	}
+	i, err = db.retrievalDataIndex.Get(item)
+	switch err {
+	case nil:
+		exists = true
+		item.StoreTimestamp = i.StoreTimestamp
+		item.BinID = i.BinID
+	case leveldb.ErrNotFound:
+		// no chunk accesses
+		exists = false
+	default:
+		return false, 0, err
+	}
+	if item.AccessTimestamp != 0 {
+		// delete current entry from the gc index
+		db.gcIndex.DeleteInBatch(batch, item)
+		gcSizeChange--
+	}
+	if item.StoreTimestamp == 0 {
+		item.StoreTimestamp = now()
+	}
+	if item.BinID == 0 {
+		item.BinID, err = db.incBinID(binIDs, db.po(item.Address))
+		if err != nil {
+			return false, 0, err
+		}
+	}
+	// update access timestamp
+	item.AccessTimestamp = now()
+	// update retrieve access index
+	db.retrievalAccessIndex.PutInBatch(batch, item)
+	// add new entry to gc index
+	db.gcIndex.PutInBatch(batch, item)
+	gcSizeChange++
+
+	db.retrievalDataIndex.PutInBatch(batch, item)
+
+	return exists, gcSizeChange, nil
```

Comment:
so any duplicates among the args only matters in this case? Are we sure? ;)

Reply:
I am not sure that I understand your comment. May it be that there is actually a problem on providing duplicate chunks in Put arguments (chunk not in the database, but will be duplicate in the batch)? This is certainly a problem that I overlooked and that it needs to be handled.